warehouse/query-core/src/main/java/datawave/query/function/LimitFields.java

```diff
@@ -87,89 +88,127 @@ public Multimap<String,Attribute<? extends Comparable<?>>> load(String key) {
+                        if (isHit(keyWithGrouping, value, hitTermMap)) {
+                            increment(keyNoGrouping, countKeepersForFieldMap);
+                        } else {
+                            value.setToKeep(false);
+                            increment(keyNoGrouping, countMissesRemainingForFieldMap);
+                        }
+                        increment(keyNoGrouping, countForFieldMap);
                     }
-                    
                 } else {
-                    manageHitsAndMisses(keyWithGrouping, keyNoGrouping, attr, hitTermMap, hits, misses, countForFieldMap);
+                    if (isHit(keyWithGrouping, attr, hitTermMap)) {
+                        increment(keyNoGrouping, countKeepersForFieldMap);
+                    } else {
+                        attr.setToKeep(false);
+                        increment(keyNoGrouping, countMissesRemainingForFieldMap);
+                    }
+                    increment(keyNoGrouping, countForFieldMap);
                 }
             }
         }
-        for (String keyNoGrouping : countForFieldMap.keySet()) {
-            int limit = this.limitFieldsMap.get(keyNoGrouping);
-            Multimap<String,Attribute<? extends Comparable<?>>> hitMap = hits.getUnchecked(keyNoGrouping);
-            for (String keyWithGrouping : hitMap.keySet()) {
-                for (Attribute<? extends Comparable<?>> value : hitMap.get(keyWithGrouping)) {
-                    // if(limit <= 0) break; // comment this line if you want to get ALL hits even if the limit is exceeded
-                    reducedMap.put(keyWithGrouping, value);
-                    limit--;
-                }
+        
+        // second pass is to set any misses back to be kept if the limit allows
+        for (Map.Entry<String,Attribute<? extends Comparable<?>>> de : document.entrySet()) {
+            String keyWithGrouping = de.getKey();
+            String keyNoGrouping = keyWithGrouping;
+            // if we have grouping context on, remove the grouping context
+            if (keyNoGrouping.indexOf('.') != -1) {
+                keyNoGrouping = keyNoGrouping.substring(0, keyNoGrouping.indexOf('.'));
             }
-            Multimap<String,Attribute<? extends Comparable<?>>> missMap = misses.getUnchecked(keyNoGrouping);
-            for (String keyWithGrouping : missMap.keySet()) {
-                for (Attribute<? extends Comparable<?>> value : missMap.get(keyWithGrouping)) {
-                    if (limit <= 0)
-                        break;
-                    reducedMap.put(keyWithGrouping, value);
-                    limit--;
+            
+            if (this.limitFieldsMap.containsKey(keyNoGrouping)) { // look for the key without the grouping context
+                int limit = this.limitFieldsMap.get(keyNoGrouping); // used below if you un-comment to get all hits
```

Comment:
The comment here is OBE'd I think. We are using the limit below.

Reply:
cool